stan/math/rev/core/operator_addition.hpp

```diff
@@ -101,12 +102,185 @@ inline var operator+(const var& a, Arith b) {
  */
 template <typename Arith, require_arithmetic_t<Arith>* = nullptr>
 inline var operator+(Arith a, const var& b) {
-  if (a == 0.0) {
-    return b;
-  }
   return b + a;  // by symmetry
 }
 
+/**
+ * Addition operator for matrix variables (C++).
+ *
+ * @tparam VarMat1 A matrix of vars or a var with an underlying matrix type.
+ * @tparam VarMat2 A matrix of vars or a var with an underlying matrix type.
+ * @param a First variable operand.
+ * @param b Second variable operand.
+ * @return Variable result of adding two variables.
+ */
+template <typename VarMat1, typename VarMat2,
+          require_all_rev_matrix_t<VarMat1, VarMat2>* = nullptr>
+inline auto add(const VarMat1& a, const VarMat2& b) {
+  check_matching_dims("add", "a", a, "b", b);
+  using op_ret_type = decltype(a.val() + b.val());
+  using ret_type = promote_var_matrix_t<op_ret_type, VarMat1, VarMat2>;
+  arena_t<VarMat1> arena_a(a);
+  arena_t<VarMat2> arena_b(b);
+  arena_t<ret_type> ret(arena_a.val() + arena_b.val());
+  reverse_pass_callback([ret, arena_a, arena_b]() mutable {
+    for (Eigen::Index i = 0; i < ret.size(); ++i) {
+      const auto ref_adj = ret.adj().coeffRef(i);
```

Comment:
Given the ```const auto``` return type is returning a template, is there a preference for doing this rather than just:
```
arena_a.adj().coeffRef(i) += ret.adj().coeffRef(i);
```
?

Reply:
Yes so this is a weird specific C++ thing. I'll pull out a godbolt tmrw but essentially just loading this return's adjoint once each loop instead of getting a reference for each accumulation makes for much nicer SIMD asm

Reply:
So essentially doing a `.coeffRef()` for the return, when working with `var<mat>`, in each accumulation causes the compiler to have to store a register for the reference and then fetch the value the reference points to for each accumulation in the loop. So since we use it more than once it's actually better to just pull out the value once and reuse it for both accumulations

In the below source 1 pull out the matrix value while source 2 uses `.coeffRef()`. You can ctrl+f for "beginnn" in the diff on the right to see the start and end of the loop code.

The main stuff is in the code that loops around like .L36 or .L35 where some of the registers we move stuff around in look a lil nicer. Lol it's very much a micro optimization but when I benchmarked this it seemed to matter


https://godbolt.org/z/W77Tj4

Reply:
> Thanks for going through the `apply_scalar_binary` benchmarking rigmarole.

That is exactly why I found this lil optim lol

Reply:
Interesting find. I guess that's like an aliasing thing. Like the compiler doesn't know the memory on the left is not the memory on the right, and so it can't actually do an optimization to avoid the two loads.

Reply:
^Yeah I think that's right